cmd/hypper/upgrade.go

```diff
@@ -0,0 +1,196 @@
+	"helm.sh/helm/v3/pkg/cli/values"
+	"helm.sh/helm/v3/pkg/getter"
+	"helm.sh/helm/v3/pkg/storage/driver"
+)
+
+const upgradeDesc = `
+This command upgrades a release to a new version of a chart.
+
+The upgrade arguments must be a release and chart. The chart
+argument can be either: a chart reference('example/mariadb'), a path to a chart directory,
+a packaged chart, or a fully qualified URL. For chart references, the latest
+version will be specified unless the '--version' flag is set.
+
+To override values in a chart, use either the '--values' flag and pass in a file
+or use the '--set' flag and pass configuration from the command line, to force string
+values, use '--set-string'. In case a value is large and therefore
+you want not to use neither '--values' nor '--set', use '--set-file' to read the
+single large value from file.
+
+You can specify the '--values'/'-f' flag multiple times. The priority will be given to the
+last (right-most) file specified. For example, if both myvalues.yaml and override.yaml
+contained a key called 'Test', the value set in override.yaml would take precedence:
+
+    $ helm upgrade -f myvalues.yaml -f override.yaml redis ./redis
+
+You can specify the '--set' flag multiple times. The priority will be given to the
+last (right-most) set specified. For example, if both 'bar' and 'newbar' values are
+set for a key called 'foo', the 'newbar' value would take precedence:
+
+    $ helm upgrade --set foo=bar --set foo=newbar redis ./redis
+`
+
+func newUpgradeCmd(cfg *action.Configuration, logger log.Logger) *cobra.Command {
+	client := action.NewUpgrade(cfg)
+	valueOpts := &values.Options{}
+	var outfmt output.Format
+	var createNamespace bool
+
+	cmd := &cobra.Command{
+		Use:   "upgrade [CHART]",
+		Short: "upgrade a release",
+		Long:  upgradeDesc,
+		Args:  require.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			client.Namespace = settings.Namespace()
+
+			wInfo := logio.NewWriter(logger, log.InfoLevel)
+
+			chartPath, err := client.ChartPathOptions.LocateChart(args[0], settings.EnvSettings)
+			if err != nil {
+				return err
+			}
+
+			// Check chart dependencies to make sure all are present in /charts
+			ch, err := loader.Load(chartPath)
+			if err != nil {
+				return err
+			}
+			if req := ch.Metadata.Dependencies; req != nil {
+				if err := action.CheckDependencies(ch, req); err != nil {
+					return err
+				}
+			}
+
+			if ch.Metadata.Deprecated {
+				logger.Warn(eyecandy.ESPrint(settings.NoEmojis, "This chart has been deprecated :exclamation:"))
+
+			}
+			if client.ReleaseName == "" {
+				client.ReleaseName, _ = client.Name(ch, args)
+				logger.Debugf("client.ReleaseName was empty, setting release name to %s", client.ReleaseName)
+			}
+
+			// Fixes #7002 - Support reading values from STDIN for `upgrade` command
+			// Must load values AFTER determining if we have to call install so that values loaded from stdin are are not read twice
```

Comment:
If I understand correctly comment specifies that the `if client.Install {}` *needs* to run before loading values. If not, values are read twice.

Which means that this block should come first thing in the `RunE: Func()` as it is in Helm.

Reply:
This actually refers to line 149 in the same file, I went through the mentioned issue and PR and the problem was that before they were loading the `vals` first and then calling `client.Install` which would load the `vals` again, thus reading the `vals` twice.

`vals, err := valueOpts.MergeValues(getter.All(settings.EnvSettings))`

Now that the `vals` are below the `client.Install` they are only read before calling `client.Upgrade`, and we leave `client.Install` to load the `vals` itself.

Reply:
Nice! many thanks for the explanation and sorry for the noise! :)